enhancements/implicit-catalog-versioning.md

```diff
@@ -0,0 +1,190 @@
+- Modify OLM to use the `spec.version` field everywhere it attempts to parse the graph for update information
+- Update skip feature to use version strings
+
+### Deprecate the `replaces` field:
+
+Currently, the update graph defined when creating an operator catalog is explicit. Essentially, the ClusterServiceVersion (CSV) file has an explicit `replaces` field:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: etcdoperator.v0.9.2
+  namespace: placeholder
+spec:
+    displayName: etcd
+    description: Etcd Operator
+    version: 0.9.2
+    replaces: etcdoperator.v0.9.0
+```
+
+The `replaces` field is the source for the update graph for each install channel in a given operator. When the `operator-registry` builds the update graph, it actually walks through each CSV in a given Operator and, for each channel, constructs a chain of CSVs that define the update path. Instead, we can just define our spec without the replaces field:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: etcdoperator.v0.9.2
+  namespace: placeholder
+spec:
+    displayName: etcd
+    description: Etcd Operator
+    version: 0.9.2
+```
+
+We would use the fact that 0.9.2 > 0.9.0 to implicitly assume that 0.9.0 is replaced by 0.9.2 through the rules of semantic version.
+
+To accomplish this:
+- Make `spec.version` a required field in the operator-registry project
+- Update the `operator-registry` to walk through the set of CSVs and construct the replacement chain by reading the `version` field instead of `replaces`
+- Add `version` as a key in the `operator-registry`'s bundle table to ensure that version has to be unique.
+- Add a `replacesVersion` field for each channel in the `operator-registry`'s database that defines which version this bundle is replacing in the update graph (along with a query to get the previous version)
+- Add a `replacedByVersion` field for each channel in the `operator-registry`'s database that defines which version this bundle is being replaced by in the update graph (along with a query to get the next version)
+- Update everywhere in OLM that references `spec.Replaces` to instead get the `replaceVersion` to determine what CSV the current is replacing *if the new fields are available*
+
+Note that these requirements will only be the case when generating update graphs with the latest version of the operator-registry's `opm registry` commands. For now, we will not update the version of the `ClusterServiceVersion` CRD type in OLM to allow for backwards compatibility
+
+### Skipping updates
+
+When OLM updates an operator,  The `spec.skips` field is used to bypass certain specific previous versions when building the update graph. This field uses the CSV `spec.name` to determine what specific CSVs to skip in the same way that the `replaces` field is used:
```

Comment:
the skips field tells operator-registry how to generate "synthetic" channel entries - my assumption is that these would now be generated implicitly from semver (i.e. 1.2.3 skips 1.2.2 and 1.2.1) because we always want users to install the latest patch (unless they have explicitly asked for an older version).

it's also a registry-only field. the skip field is not used on-cluster at all, and isn't in the csv schema.

Reply:
I wasn't intending for us to make assumptions about patch version in particular. I think if that's a rule that OpenShift in particular wants to enforce, they can just do it through release pipelines doing some linting. It seems possible that a user might want to manage whether patch versions are safe to skip over and which aren't.